src/rust/inetstack/protocols/tcp/tests/simulator.rs

```diff
@@ -0,0 +1,827 @@
+    runtime::{
+        memory::DemiBuffer,
+        network::PacketBuf,
+    },
+    QDesc,
+};
+
+//======================================================================================================================
+// Standalone Functions
+//======================================================================================================================
+
+#[test]
+/// Runs the network test suite.
+fn test_simulation() -> Result<()> {
+    let verbose: bool = false;
+    let local_mac: MacAddress = test_helpers::ALICE_MAC;
+    let remote_mac: MacAddress = test_helpers::BOB_MAC;
+    let local_sockaddr: SocketAddrV4 = SocketAddrV4::new(test_helpers::ALICE_IPV4, 80);
+    let remote_sockaddr: SocketAddrV4 = SocketAddrV4::new(test_helpers::BOB_IPV4, 49751);
+
+    let input_path: String = match env::var("INPUT_DIR") {
+        Ok(config_path) => config_path,
+        Err(e) => {
+            let cause: String = format!("missing INPUT_DIR environment variable (err={:?})", e);
+            eprintln!("test_simulation(): {:?}", cause);
+            anyhow::bail!(cause);
+        },
+    };
+
+    let tests: Vec<String> = collect_tests(&input_path)?;
+
+    if tests.is_empty() {
+        let cause: String = format!("no tests found under {:?}", input_path);
+        eprintln!("test_simulation(): {:?}", cause);
+        return Ok(());
+    }
+
+    for test in &tests {
+        println!("Running test: {:?}", test);
+
+        let mut simulation: Simulation =
+            Simulation::new(test, &local_mac, &local_sockaddr, &remote_mac, &remote_sockaddr)?;
+        simulation.run(verbose)?;
+    }
+
+    Ok(())
+}
+
+// Collect all files under 'test_path'.
+fn collect_tests(test_path: &str) -> Result<Vec<String>> {
+    let mut files: Vec<String> = Vec::new();
+    for entry in std::fs::read_dir(test_path)? {
+        let entry: DirEntry = entry?;
+        // Skip directories.
+        if entry.file_type()?.is_dir() {
+            continue;
+        }
+
+        // Skip unsupported files.
+        if !entry.file_name().to_str().unwrap().ends_with(".pkt") {
+            continue;
+        }
+
+        let path: PathBuf = entry.path();
+        let path: String = path.to_str().unwrap().to_string();
+        files.push(path);
+    }
+
+    files.sort();
+
+    Ok(files)
+}
+
+//======================================================================================================================
+// Standalone Functions
+//======================================================================================================================
+
+/// A simulation of the network stack.
+struct Simulation {
+    local_mac: MacAddress,
+    local_addr: SocketAddrV4,
+    remote_mac: MacAddress,
+    remote_addr: SocketAddrV4,
+    local_qd: Option<(u32, QDesc)>,
+    remote_qd: Option<(u32, Option<QDesc>)>,
+    server: SharedEngine<RECEIVE_BATCH_SIZE>,
+    now: Instant,
+    inflight: Option<QToken>,
+    steps: Vec<String>,
+}
+
+impl Simulation {
```

Comment:
Suggestion: This file/class and some of the functions are long. Is it possible to split it into smaller groups of functionalities? If we don't expect much change in this class then it's not a big concern.

Reply:
I intend to be working on this whole module for the next months. As we agreed, I just pushed an initial version so that it does not break recurrently with all the changes that are being introduced.

Reply:
Yes, that sounds good.